File: search/binarysearch.py

max = 0
def linearSearch(arr, key):
   n = len(arr)
   for i in range(0, n):
      if arr[i] == key:
         return i
   return -1

def binarySearch(arr, item):
    first = 0
    last = len(arr)-1
    found = False
    count = 0
    global max
    while first<=last and not found:
        midpoint = (first + last)//2
        count = count + 1
        if arr[midpoint] == item:
            found = True
            if max <count:
               max = count
            #print ("found  " , item, "comp  = ", count, " max = ", max)
            count = 0
        else:
            if item < arr[midpoint]:
                last = midpoint-1
            else:
                first = midpoint+1

    return found

arr = list(range(0,1001))

for i in range(0,1000):
    binarySearch(arr, i)

list = [1,2,3,4,5]
key = 2

pos = binarySearch(list, key)
print ("position = ", pos )
if pos == -1:
  print("not found" )
else:
  print (" ",  list[pos])

key = -1
pos = binarySearch(list, key)
print ("position = ", pos )
if pos == -1:
  print(" not found" )
else:
  print ( list[pos])


if __name__ == '__main__':
    import timeit
    import random
    t1 = timeit.Timer("list=range(100000); k = random.randint(1,100000); key=k; linearSearch(list,key)", "from __main__ import linearSearch, random")
    print("linearSearch search :", t1.timeit(number=100), "milliseconds")

    t1 = timeit.Timer("list=range(1000000); k = random.randint(1,1000000); key=k; binarySearch(list,key)", "from __main__ import binarySearch, random")
    print("binary search :", t1.timeit(number=100), "milliseconds")

File: 02-analysis/search/BinarySearch.java
// 프로그램 2.3: 이진 탐색
public class BinarySearch {

	public static void main(String[] args) {
		int[] A = {10, 12, 13, 14, 18, 20, 25, 27, 30, 35, 40, 45, 47};
		int x = 18;
		int n = A.length;
		int location;

		location = binarySearch(A, 0, n - 1, x);

		System.out.println(x + "의 지수 = " + location);
	}

	// target이 배열 A[first .. last]내에 있는지를 확인하기 위해 이진 탐색을 사용한다
	// 매개 변수: 배열 A[first .. last], target(찾는 값)
	// 반환값: 있으면 해당 요소의 지수, 없으면 -1
	public static int binarySearch(int[] A, int first, int last, int target)
    {
        int result;

        if (first > last) // 찾지 못한 경우
            result = -1;
        else
        {
	        // 중간 요소의 지수를 구한다
            int mid = (first + last) / 2;

            if (target == A[mid]) // target이 중간 요소와 같아 찾은 경우
                result = mid;
            else if (target < A[mid]) // target이 중간 요소보다 작다면 앞쪽 반에서 찾느다
                result = binarySearch(A, first, mid - 1, target);
            else // target이 중간 요소보다 크다면 뒤쪽 반에서 찾느다
                result = binarySearch(A, mid + 1, last, target);
        }
        return result;
    }
}